File: codejam/2008/round_2/triangle_areas/solution.hs
import Data.List (intersperse)

infixr 0 <|
(<|) f x = f x
(<||) f g x = f <| g x

triangleCoordinates :: [Int] -> Maybe ((Int, Int), (Int, Int))
triangleCoordinates [n, m, a]
    | a > n*m      = Nothing
    | mod a n == 0 = Just ((n, 0), (0, div a n))
    | otherwise    = Just ((x0, y0), (1, y1))
    where
        y1 = 1 + div a n
        x0 = 1 + div a y1
        y0 = y1*x0 - a

format :: Maybe ((Int, Int), (Int, Int)) -> String
format Nothing = "IMPOSSIBLE"
format (Just ((x0, y0), (x1, y1))) = "0 0 "
    ++ (concat
        <| intersperse " "
        <| map show [x0, y0, x1, y1])

solve :: IO String
solve = getLine
    >>= return
        <|| format
        <|| triangleCoordinates
        <|| map (read :: String -> Int)
        <|| words

main :: IO ()
main = (readLn :: IO Int)
    >>= sequence <|| flip replicate solve
    >>= putStrLn
        <|| concat
        <|| intersperse "\n"
        <|| map (uncurry (++))
        <|| zip (map (("Case #" ++) <|| (++ ": ") <|| show) [1..])

File: codejam/2008/round_2/triangle_areas/solution.rs
lib::run!();

struct TestCase {
    n: usize,
    m: usize,
    a: usize,
}

fn read() -> TestCase {
    lib::input!([usize; 3] as [n, m, a]);
    TestCase { n, m, a }
}

fn solve(TestCase { n, m, a }: TestCase) -> String {
    if n * m < a {
        "IMPOSSIBLE".to_string()
    } else {
        let [x1, y1, x2, y2]: [usize; 4];
        if a % n == 0 {
            y2 = a / n;
            x1 = n;
            x2 = 0;
            y1 = 0;
        } else {
            y2 = 1 + a / n;
            x1 = 1 + a / y2;
            x2 = 1;
            y1 = (x1 * y2) - a;
        }
        assert!(x1 * y2 - x2 * y1 == a);
        assert!(x1 <= n);
        assert!(x2 <= n);
        assert!(y1 <= m);
        assert!(y2 <= m);
        format!("0 0 {x1} {y1} {x2} {y2}")
    }
}
